packages/vscode-extension/src/panels/WorkspaceConfigController.ts

```diff
@@ -1,86 +1,73 @@
-        showDeviceFrame: config.get<boolean>("showDeviceFrame")!,
-        stopPreviousDevices: config.get<boolean>("stopPreviousDevices")!,
-      };
+    this.disposables.push(
+      workspace.onDidChangeConfiguration((event: ConfigurationChangeEvent) => {
+        if (!event.affectsConfiguration("RadonIDE")) {
+          return;
+        }
+        const config = workspace.getConfiguration("RadonIDE");
 
-      if (newConfig.panelLocation !== this.config.panelLocation) {
-        getTelemetryReporter().sendTelemetryEvent(
-          "workspace-configuration:panel-location-changed",
-          { newPanelLocation: newConfig.panelLocation }
-        );
-      }
+        const newConfig = {
+          panelLocation: config.get<PanelLocation>("panelLocation")!,
+          showDeviceFrame: config.get<boolean>("showDeviceFrame")!,
+          stopPreviousDevices: config.get<boolean>("stopPreviousDevices")!,
+        };
 
-      if (newConfig.showDeviceFrame !== this.config.showDeviceFrame) {
-        getTelemetryReporter().sendTelemetryEvent(
-          "workspace-configuration:show-device-frame-changed",
-          { showDeviceFrame: String(newConfig.showDeviceFrame) }
-        );
-      }
+        const oldConfig = this.stateManager.getState();
 
-      this.config = newConfig;
-      this.eventEmitter.emit("configChange", this.config);
-    });
-  }
+        if (newConfig.panelLocation !== oldConfig.panelLocation) {
+          getTelemetryReporter().sendTelemetryEvent(
+            "workspace-configuration:panel-location-changed",
+            { newPanelLocation: newConfig.panelLocation }
+          );
+        }
 
-  async getConfig() {
-    return this.config;
-  }
+        if (newConfig.showDeviceFrame !== oldConfig.showDeviceFrame) {
+          getTelemetryReporter().sendTelemetryEvent(
+            "workspace-configuration:show-device-frame-changed",
+            { showDeviceFrame: String(newConfig.showDeviceFrame) }
+          );
+        }
 
-  async update<K extends keyof WorkspaceConfigProps>(key: K, value: WorkspaceConfigProps[K]) {
-    const configuration = workspace.getConfiguration("RadonIDE");
-    if (configuration.inspect(key as string)?.workspaceValue) {
-      await configuration.update(key as string, value, false);
-    } else {
-      await configuration.update(key as string, value, true);
-    }
-  }
+        this.stateManager.setState(newConfig);
+      })
+    );
 
-  async addListener<K extends keyof WorkspaceConfigEventMap>(
-    eventType: K,
-    listener: WorkspaceConfigEventListener<WorkspaceConfigEventMap[K]>
-  ) {
-    this.eventEmitter.addListener(eventType, listener);
-  }
+    this.stateManager.on("setState", async (partialState) => {
+      const partialStateEntries = Object.entries(partialState);
+
+      const config = workspace.getConfiguration("RadonIDE");
+
+      for (const partialStateEntry of partialStateEntries) {
+        if (config.inspect(partialStateEntry[0] as string)?.workspaceValue) {
```

Comment:
```suggestion
        if (config.inspect(partialStateEntry[0] as string)?.workspaceValue === undefined) {
```
Do we want this branch on falsy values as well?

Reply:
yes this is a filed that informs us if this is a workspace or user setting